src/meta-client/src/client/ddl.rs

```diff
@@ -133,23 +135,41 @@ impl Inner {
         loop {
             if let Some(leader) = &ask_leader.get_leader() {
                 let mut client = self.make_client(leader)?;
-                let res = client
+
+                match client
                     .submit_ddl_task(req.clone())
                     .await
-                    .context(error::TonicStatusSnafu)?;
-
-                let res = res.into_inner();
-
-                if let Some(header) = res.header.as_ref() {
-                    if let Some(err) = header.error.as_ref() {
-                        if err.code == ErrorCode::NotLeader as i32 {
-                            let _ = ask_leader.ask_leader().await?;
+                    .context(error::TonicStatusSnafu)
+                {
+                    Ok(res) => {
+                        let res = res.into_inner();
+
+                        if let Some(header) = res.header.as_ref() {
+                            if let Some(err) = header.error.as_ref() {
+                                if err.code == ErrorCode::NotLeader as i32 {
+                                    let _ = ask_leader.ask_leader().await?;
+                                    continue;
+                                }
+                            }
+                        }
+                        return Ok(res);
+                    }
+                    // The leader may be unreachable.
+                    Err(err) => {
+                        let error::Error::TonicStatus { source, .. } = &err else {
+                            unreachable!()
+                        };
+
+                        if source.code() == Code::Unavailable {
+                            warn!("Submitting ddl to {leader} is unreachable, try to update leader addr");
+                            let leader = ask_leader.ask_leader().await?;
+                            info!("DDL client updated to new leader addr: {leader}");
                             continue;
+                        } else {
+                            return Err(err);
```

Comment:
Here it feels like there is room for further optimization. I suggest that regardless of any error, we should blindly retry calling `ask_leader`. Additionally, in scenarios where domain names are used, there is a probability of failure when calling `ask_leader` (as it may route to a node that is down). Therefore, we cannot use '?' for ask_leader. It might also be necessary to retry ask_leader multiple times. I recommend retrying up to 3 times and if it still fails, then throw an error and exit the loop.